### PR TITLE
Refactor Code Coverage Support

### LINKED ISSUE #274: Make Jacoco and Cobertura plugins no longer required
Have the ability to parse and report coverage from cobertura/jacoco xml reports natively in the phabricator jenkins plugin. Currently only cobertura is supported (in the release version of the plugin). Even when the cobertura plugin is applied, we still process all the coverage xml files to collect line coverage. This change will completely remove any reliance on external coverage jenkins plugins and process coverage xml for cobertura and jacoco natively.

Since the jacoco coverage currently available on master has not be released yet, it will be refactored to remove the bits that are reliant on the jacoco plugin and the parsing of the jacoco exec binary format

### PR DESCRIPTION
**Pull request by kageiit**
- Make Jacoco and Cobertura plugins no longer required
- Coverage is processed via xml reports supplied by cobertura/jacoco
- Completely decouple coverage parsing logic from jenkins/hudson dependencies so they can be used independently
- No longer supporting overwritten coverage files (https://github.com/uber/phabricator-jenkins-plugin/pull/167). Only agrgegation of coverage is supported. If there are duplicate coverage elements across multiple files, the behavior is to pick the last available coverage value.
- Fixes #274

### CODE REVIEW

**Comment by kageiit** [comment]
cc @ChaitanyaPramod

File: src/main/java/com/uber/jenkins/phabricator/coverage/XmlCoverageProvider.java [hunk 1]
```diff
@@ -0,0 +1,408 @@
...
+                lineCoverage.put(entry.getKey(), sortedCounts);
+            }
+        }
+
+        static int getIntValue(Node node, String attributeName) {
+            return Integer.parseInt(node.getAttributes().getNamedItem(attributeName).getTextContent());
+        }
+    }
+
+    private static class CoberturaXmlCoverageHandler extends XmlCoverageHandler {
```
Comment (nguyentruongtho) [inline src/main/java/com/uber/jenkins/phabricator/coverage/XmlCoverageProvider.java]: I assume below handlers are brought over from corresponding jenkins plugin, should we add the licenses for those?

File: src/main/java/com/uber/jenkins/phabricator/PhabricatorNotifier.java [hunk 2]
```diff
@@ -319,36 +327,55 @@ private CoverageProvider getCoverageProvider(
...
+            report.delete();
         }
+    }
 
-        logger.info(UBERALLS_TAG, "No coverage results found");
-        return null;
+    private Set<File> getCoverageReports(Run<?, ?> build) {
+        Set<File> reports = new HashSet<>();
+
+        File[] foundReports = build.getRootDir().listFiles(COVERAGE_FILENAME_FILTER);
```
Comment (nguyentruongtho) [inline src/main/java/com/uber/jenkins/phabricator/PhabricatorNotifier.java]: should we cleanup jacoco and cobertura reports as well?

File: src/main/java/com/uber/jenkins/phabricator/coverage/XmlCoverageProvider.java (hunk 1 above)
Comment (kageiit) [reply src/main/java/com/uber/jenkins/phabricator/coverage/XmlCoverageProvider.java]: This was written as a clean room implementation and does not follow the upstream version at all

File: src/main/java/com/uber/jenkins/phabricator/PhabricatorNotifier.java (hunk 2 above)
Comment (kageiit) [reply src/main/java/com/uber/jenkins/phabricator/PhabricatorNotifier.java]: The files are always renamed to `coverage*.xml` during the copy step

Comment (nguyentruongtho) [reply src/main/java/com/uber/jenkins/phabricator/PhabricatorNotifier.java]: right, my bad.

**Comment by kageiit** [comment]
Ran a full integration test as follows
- Using a real phabricator instance, ran a diff job on the local server spun up via `./gradlew server`
- Planted coverage files (both jacoco and cobertura) from a real diff into the local server job's workspace so the plugin could pick them up
- Verified that the coverage inline etc. are exactly as expected before and after for both cobertura and jacoco
- Also verified that coverage shows correctly now for kotlin classes instrumented via jacoco

**Pull request merged**